src/layers/domain/core/questionnaire.py

```diff
@@ -27,49 +30,238 @@ class Question(BaseModel):
     name: str
     type: QuestionType
     multiple: bool
+    validation_rules: list[str] = []
+    choices: list[str] = []
 
 
 class Questionnaire(BaseModel):
     """
     A Questionnaire represents a collection of Questions, in a specific order.
     """
 
-    id: UUID
     name: str = Field(regex=ENTITY_NAME_REGEX)
-    _questions: list[Question] = (lambda: [])()
+    version: int
+    questions: list[Question] = []
 
     def __contains__(self, question_name: str) -> bool:
         """
         Returns true if the question specified exists within the questionnaire
         """
-        return question_name in (q.name for q in self._questions)
+
+        return question_name in (q.name for q in self.questions)
 
     def add_question(
         self,
         name: str,
         type: QuestionType = QuestionType.STRING,
         multiple: bool = False,
+        validation_rules: list[str] = [],
+        choices: list[str] = [],
     ):
         """
-        Adds a new question to the questionnaire.  Once a questionnaire is
-        locked this method will throw an error.
+        Adds a new question to the questionnaire.
         """
-        if name in self:
-            raise DuplicateError(f"Question exists: {name}")
-        result = Question(name=name, type=type, multiple=multiple)
-        self._questions.append(result)
-        return result
+        for question in self.questions:
+            if name == question.name:
+                raise DuplicateError(f"Question exists: {name}")
 
-    def remove_question(self, name: str):
-        """
-        Removes a question from the questionnaire.  Once a questionnaire is
-        locked this method will throw an error.
-        """
-        raise NotImplementedError()
+        question = Question(
+            name=name,
+            type=type,
+            multiple=multiple,
+            validation_rules=validation_rules,
+            choices=choices,
+        )
+        self.questions.append(question)
+        return question
 
-    def revise(self) -> "Questionnaire":
-        """
-        Creates an exact duplicate of the questionnaire, which is the only way
-        to edit questionnaires once they become locked.
-        """
-        raise NotImplementedError()
+
+class QuestionnaireResponse(BaseModel):
+    """
+    Represents a Questionnaire response mapping Questions to Responses.
+    """
+
+    responses: dict
+
+
+class QuestionnaireResponseValidator(BaseModel):
+    """
+    Validates questionnaire responses against questionnaire questions are of the correct type, meet any validation criteria and .
+    """
+
+    questionnaire: Questionnaire
+    responses: dict
```

Comment:
if we group all responses by question at the start then we can reduce the code later on, so I'd suggest `responses: list[tuple[str, list]]` (i.e. a list of question -> answers) for two reasons

* if we make the `responses` object a `list` then we can leverage pydantics `@validator(..., each_item=True)` to validate each response separately
* if we make all question answers `list`, even if multiple=False - this will allow us to simplify all downstream code as we won't need the `list`/not `list` branching